apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/TraceContext.java

```diff
@@ -428,10 +464,51 @@ String getIncomingTraceParentHeader() {
         return sb.toString();
     }
 
+    /**
+     * Sets Trace context text headers, using this context as parent, on the provided carrier using the provided setter
+     *
+     * @param carrier      the text headers carrier
+     * @param headerSetter a setter implementing the actual addition of headers to the headers carrier
+     * @param <C>          the header carrier type, for example - an HTTP request
+     */
+    public <C> void setOutgoingTraceContextHeaders(@Nullable C carrier, TextHeaderSetter<C> headerSetter) {
+        if (carrier == null) {
+            logger.warn("Attempted to set a text header to a null carrier through {}. Distributed tracing cannot be supported.", headerSetter.getClass().getName());
+            return;
+        }
+        headerSetter.setHeader(TRACE_PARENT_TEXTUAL_HEADER_NAME, getOutgoingTraceParentTextHeader().toString(), carrier);
+    }
+
+    /**
+     * Sets Trace context binary headers, using this context as parent, on the provided carrier using the provided setter
+     *
+     * @param carrier      the binary headers carrier
+     * @param headerSetter a setter implementing the actual addition of headers to the headers carrier
+     * @param <C>          the header carrier type, for example - a Kafka record
+     * @return true if Trace Context headers were set; false otherwise
+     */
+    public <C> boolean setOutgoingTraceContextHeaders(@Nullable C carrier, BinaryHeaderSetter<C> headerSetter) {
+        if (carrier == null) {
+            logger.warn("Attempted to set a binary header to a null carrier through {}. Distributed tracing cannot be supported.", headerSetter.getClass().getName());
+            return false;
+        }
+        byte[] buffer = headerSetter.getFixedLengthByteArray(TRACE_PARENT_BINARY_HEADER_NAME, BINARY_FORMAT_EXPECTED_LENGTH);
+        if (buffer == null || buffer.length != BINARY_FORMAT_EXPECTED_LENGTH) {
+            logger.warn("Header setter {} failed to provide a byte buffer with the proper length. Allocating a buffer for each header.",
```

Comment:
There is a single implementation (Kafka) that returns `null` if the current assumption (of same thread setting and serializing the headers) becomes invalid. If and when that happens, I don't mind the first user encountering it will get lots of errors. I made it so that distributed tracing will continue to work, but will start allocating more than before, so the can be easily overlooked if not logged clearly.

Reply:
We should track https://issues.apache.org/jira/browse/KAFKA-3539